homeassistant/components/openuv/__init__.py

```diff
@@ -82,33 +160,76 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
 
     await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
 
+    @callback
+    def extract_openuv(func: Callable) -> Callable:
+        """Define a decorator to get the correct OpenUV object for a service call."""
+
+        async def wrapper(call: ServiceCall) -> None:
+            """Wrap the service function."""
+            openuv = async_get_openuv_for_service_call(hass, call)
+
+            try:
+                await func(call, openuv)
+            except OpenUvError as err:
+                raise HomeAssistantError(
+                    f'Error while executing "{call.service}": {err}'
+                ) from err
+
+        return wrapper
+
     @_verify_domain_control
-    async def update_data(_: ServiceCall) -> None:
+    @extract_openuv
+    async def update_data(call: ServiceCall, openuv: OpenUV) -> None:
         """Refresh all OpenUV data."""
         LOGGER.debug("Refreshing all OpenUV data")
+        async_log_deprecated_service_call(
+            hass,
+            call,
+            "homeassistant.update_entity",
+            "binary_sensor.protection_window, sensor.current_uv_index",
```

Comment:
Is this the correct entity id if they have multiple?

Reply:
Roger – the string that this populates is here: https://github.com/home-assistant/core/blob/e4490145c7f2b77f05389148b89a9710c8fba386/homeassistant/components/openuv/strings.json#L39

The idea here is that, in this case, either entity will update the relevant data. If you think the language isn't clear enough, open to ideas. 👍🏻

Reply:
Can they have multiple openuv integrations , or have renamed the entity_ids ?

Reply:
Hmm, good point. I probably need to fish the correct entity IDs out of the registry based on which config entry we're looking at.

Reply:
Latest flow committed. First, we call the service with a config entry selector:

![CleanShot 2022-09-17 at 16 59 59](https://user-images.githubusercontent.com/47216/190878993-3683d9cb-1030-430c-a2fa-ba20ae1af7b8.png)

When we call the service, we get this log message:

```
2022-09-17 16:46:19.168 WARNING (MainThread) [homeassistant.components.openuv] The "openuv.update_data" service is deprecated and will be removed in 2022.12.0; review the Repairs item in the UI for more information
```

...and this Repair entry:

![CleanShot 2022-09-17 at 16 59 35](https://user-images.githubusercontent.com/47216/190878988-26e0cad3-41af-48fa-bbbe-da389b2f38e9.png)

![CleanShot 2022-09-17 at 17 02 31](https://user-images.githubusercontent.com/47216/190879037-ec1fb9c1-449b-4ac5-8cb9-93c56c1f54f2.png)